wc.rbを追加

wak9102 opened the pull request:
wc コマンドを作るの課題提出のため、06.wcにwc.rbを追加しました

wak9102 committed:
first commit

wak9102 committed:
06.wcにwc.rbを作成、カウント機能を追加

課題提出用にwc.rbを作成しました。
ファイルおよび標準入力から行数、文字数、バイト数をカウントする機能を追加しました。

wak9102 committed:
wc.rbにカウント結果の出力機能を追加

取得したカウント結果をオプションと引数の数に応じて分岐して出力する機能を追加しました。

wak9102 committed:
wc.rbをrubocopにより修正

rubocopの検出にそってメソッド等を変更しました

wak9102 committed:
wc.rbの誤字を修正

optionの要素を:linesではなく:lineと誤記載していた部分を修正しました

wak9102 committed:
wc.rbの一部をリファクタリング

hashの新規作成、unless修飾子の箇所を変更

wak9102 committed:
wc.rbにmainメソッドを追加

コードの読みやすさとrubocopによるチェック精度を上げるため、共通の処理をmainメソッドとしてまとめました

wak9102 committed:
wc.rbでrubocopの指摘箇所を修正

IfUnlessModifier、GuardClauseに関する指摘箇所を修正しました

wak9102 committed:
wc.rbの各処理をメソッド化

mainメソッドの行数と複雑性を削減するため、カウント処理などをメソッドに分けました

wak9102 committed:
wc.rbの誤字を修正

呼び出すメソッド名に誤字があったので修正しました

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,61 @@
+def wc_count_stdin
+  counted_result = { name: :stdin }
+  stdin_text = $stdin.readlines.join
+  counted_result.merge!(wc_count_text(stdin_text))
```
ここは `merge!` ではなく `merge` でも良いですね。
破壊的メソッドは使わなくて良いならそれに越したことはないので。

wak9102 replied:
@yuuu さん
ご教示ありがとうございます。mergeに変更完了いたしました。

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,61 @@
+end
+
+def counted_results_total_print(counted_results, options)
+  print counted_results.inject(0) { |sum, result| sum + result[:line] }.to_s.rjust(8)
```
ここは `Array#inject` よりも `Array#sum` の方が適任ですね。

```ruby
# injectの場合
counted_results.inject(0) { |sum, result| sum + result[:line] }

# sumの場合
counted_results.sum { |result| result[:line] }
```

`Array#inject` を使うのはもう少し複雑な計算や集計を行うときですね。

wak9102 replied:
@yuuu さん
ご教示ありがとうございます。sumを使用する形に変更しました。
（こちら以前にも同様につまづいた箇所でした💦　改めてリファレンスを読み直しておきます）

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,61 @@
+def counted_results_print(counted_results, options)
+  counted_results.each do |result|
+    print result[:line].to_s.rjust(8)
+    print result[:word].to_s.rjust(8) if options.none?(:lines)
```
2回同じif文を書くなら、まとめたほうが読みやすいかもと思いました。
ただ、ここは人によって意見が分かれそうなので、修正は不要です。

```ruby
if options.none?(:lines)
  print result[:word].to_s.rjust(8)
  print result[:byte].to_s.rjust(8)
end
```

wak9102 replied:
@yuuu さん
ご教示ありがとうございます。確かに自身もここはどうすべきか迷いました。前後の文脈なども見つつ選択するようにしていきます。

yuuu commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,61 @@
+  # コマンドライン引数ごとのワードカウント結果を配列に格納
+  # コマンドライン引数がない場合は標準入力を参照して結果を格納
+  counted_results = ARGV.map { |file| wc_count_file(file) }
+  counted_results << wc_count_stdin if ARGV.none?
```
標準入力と引数が両方存在する場合、本物の `wc` コマンドは引数を優先するのですが、この実装だと両方がカウント対象となりますね。 せっかくなので、 `wc` コマンドに合わせた挙動に修正してもらっても良いでしょうか？

wak9102 replied:
@yuuu さん
ご教示ありがとうございます。引数が存在する場合のみ、wc_count_fileメソッドを実行するように変更いたしました。

wak9102 committed:
wc.rbをリファクタリング

wc_count_fileメソッドとwc_count_stdinメソッドのmerge!をmergeに変更
counted_results_total_printメソッドのinjectをsumを使用したコードに変更

wak9102 committed:
wc.rbで配列counted_resultsの格納手順を変更

引数が無い場合はwc_count_fileメソッドを呼び出さないように変更
